File: extensions/sdk/node/configuration-controller/test/controller.test.mjs
import assert from "node:assert/strict";
import test from "node:test";

import {
  configurationControllerCapability,
  configurationControllerContribution,
  configurationDigest,
  mergeManagedCredentials,
  replacedManagedCredentials,
  validateObservation,
  prepareRequestDigest,
} from "../src/index.mjs";

const request = {
  candidateId: `pcfg_${"a".repeat(32)}`,
  configurationId: `cfg_${"9".repeat(24)}`,
  catalogDigest: "c".repeat(64), schemaDigest: "d".repeat(64), artifactSha256: "e".repeat(64),
  expectedActive: { revision: 1, digest: "f".repeat(64) }, values: { capacity: 10 },
  managedCredentials: { token: { handle: "credential://managed/opaque", scope: "tenant", owner: "cn.vastplan.demo", purpose: "demo.token", version: 1 } },
};

test("derives an opaque controller capability and stable normalized digests", () => {
  assert.equal(configurationControllerCapability("cn.vastplan.example.hot-controller"), "configuration.3c183e12decc8e57e3ea513837dc8708");
  assert.equal(prepareRequestDigest(request), "1f14aa9cf75025b0480065230c7ae1c34f81117072b846a07b8619df6323c744");
  assert.equal(configurationDigest(request.values, request.managedCredentials), "0f4e0e9504882ed26dffe20c7d6cb101c7a4178dc73d7f58b1d1fd0a59d40210");
  assert.equal(prepareRequestDigest({ ...request, values: { z: true, capacity: 10 } }), prepareRequestDigest({ ...request, values: { capacity: 10, z: true } }));
  assert.equal(configurationDigest({ b: 2, a: 1 }), configurationDigest({ a: 1, b: 2 }));
});

test("retains omitted managed credentials and selects only replaced refs for retirement", () => {
  const retained = { handle: "credential://managed/retained", scope: "tenant", owner: "cn.vastplan.demo", purpose: "demo.token", version: 1 };
  const oldRef = { ...retained, handle: "credential://managed/old" };
  const newRef = { ...retained, handle: "credential://managed/new", version: 2 };
  const merged = mergeManagedCredentials({ retained, token: oldRef }, { token: newRef });
  assert.equal(merged.retained.handle, retained.handle);
  assert.equal(merged.token.version, 2);
  assert.deepEqual(replacedManagedCredentials({ retained, token: oldRef }, merged), [oldRef]);
});

test("requires an RFC 3339 observation timestamp", () => {
  const base = {
    protocol: "configuration.v1",
    configurationId: request.configurationId,
    active: request.expectedActive,
    observedAt: "2026-07-23T00:00:00Z",
  };
  assert.equal(validateObservation(base).observedAt, base.observedAt);
  assert.throws(() => validateObservation({ ...base, observedAt: "2026-07-23" }), /observation/);
});

test("contribution rejects direct users and returns digest-only observations", async () => {
  const controller = {
    async prepare(value) {
      return {
        protocol: "configuration.v1", configurationId: value.configurationId, active: value.expectedActive,
        candidate: { candidateId: value.candidateId, requestDigest: prepareRequestDigest(value), configurationDigest: configurationDigest(value.values, value.managedCredentials), status: "Prepared", ready: true },
        observedAt: "2026-07-23T00:00:00Z",
      };
    },
    async commit() { throw new Error("unused"); }, async abort() { throw new Error("unused"); }, async status() { throw new Error("unused"); },
  };
  const contribution = configurationControllerContribution("cn.vastplan.example.hot-controller", controller);
  const invocation = { throwIfCancelled() {} };
  const denied = await contribution.handlers.get("prepare")(invocation, {}, { tenant_id: "tenant-a", caller: { kind: "CALLER_KIND_USER", id: "alice" } }, Buffer.from(JSON.stringify(request)));
  assert.equal(denied.result.error.code, "configuration.controller.permission_denied");
  const accepted = await contribution.handlers.get("prepare")(invocation, {}, { tenant_id: "tenant-a", caller: { kind: "CALLER_KIND_PLUGIN", id: "cn.vastplan.platform.configuration.plugin-settings" } }, Buffer.from(JSON.stringify(request)));
  assert.equal(accepted.result.status, "STATUS_OK");
  const observation = JSON.parse(accepted.payload);
  assert.equal(observation.candidate.status, "Prepared");
  assert.equal(JSON.stringify(observation).includes("credential://"), false);
});

File: extensions/sdk/go/configurationcontroller/controller_test.go
package configurationcontroller

import (
	"context"
	"encoding/json"
	"strings"
	"testing"
	"time"

	configurationv1 "cdsoft.com.cn/VastPlan/contracts/schemas/configuration/v1"
	pluginv1 "cdsoft.com.cn/VastPlan/contracts/schemas/plugin/v1"
	contractv1 "cdsoft.com.cn/VastPlan/core/shared/go/contract/v1"
	sdk "cdsoft.com.cn/VastPlan/extensions/sdk/go/plugin"
)

type fakeController struct{}

func (fakeController) Prepare(_ context.Context, _ sdk.Host, _ *contractv1.CallContext, request configurationv1.PrepareRequest) (configurationv1.Observation, error) {
	digest, _ := configurationv1.DigestPrepareRequest(request)
	configurationDigest, _ := configurationv1.DigestConfiguration(request.Values, request.ManagedCredentials)
	return configurationv1.Observation{
		Protocol: configurationv1.Protocol, ConfigurationID: request.ConfigurationID, Active: request.ExpectedActive,
		Candidate:  &configurationv1.CandidateObservation{CandidateID: request.CandidateID, RequestDigest: digest, ConfigurationDigest: configurationDigest, Status: configurationv1.StatusPrepared, Ready: true},
		ObservedAt: time.Now().UTC(),
	}, nil
}

func (fakeController) Commit(context.Context, sdk.Host, *contractv1.CallContext, configurationv1.CandidateRequest) (configurationv1.Observation, error) {
	return configurationv1.Observation{}, nil
}
func (fakeController) Abort(context.Context, sdk.Host, *contractv1.CallContext, configurationv1.CandidateRequest) (configurationv1.Observation, error) {
	return configurationv1.Observation{}, nil
}
func (fakeController) Status(context.Context, sdk.Host, *contractv1.CallContext, configurationv1.StatusRequest) (configurationv1.Observation, error) {
	return configurationv1.Observation{}, nil
}

func TestContributionUsesOpaqueCapabilityAndExactCoordinator(t *testing.T) {
	const pluginID = "cn.vastplan.example.hot-controller"
	contribution, err := Contribution(pluginID, fakeController{})
	if err != nil {
		t.Fatal(err)
	}
	want, _ := pluginv1.ConfigurationControllerCapability(pluginID)
	if contribution.ExtensionPoint != configurationv1.ExtensionPoint || contribution.ID != want || strings.Contains(contribution.ID, pluginID) {
		t.Fatalf("controller contribution 身份错误: %+v", contribution)
	}
	request := configurationv1.PrepareRequest{
		CandidateID: "pcfg_" + strings.Repeat("a", 32), ConfigurationID: "cfg_" + strings.Repeat("b", 24),
		CatalogDigest: strings.Repeat("c", 64), SchemaDigest: strings.Repeat("d", 64), ArtifactSHA256: strings.Repeat("e", 64),
		ExpectedActive: configurationv1.ActiveReference{Revision: 1, Digest: strings.Repeat("f", 64)}, Values: json.RawMessage(`{"enabled":true}`),
	}
	payload, _ := json.Marshal(request)
	operation := configurationv1.OperationPrepare
	handler := contribution.Handlers[operation]
	directUser := &contractv1.CallContext{TenantId: "tenant-a", Caller: &contractv1.Caller{Kind: contractv1.CallerKind_CALLER_KIND_USER, Id: "alice"}}
	result, _, err := handler(context.Background(), nil, directUser, payload)
	if err != nil || result.GetError().GetCode() != "configuration.controller.permission_denied" {
		t.Fatalf("浏览器用户不得直接调用 controller: result=%+v err=%v", result, err)
	}
	coordinator := &contractv1.CallContext{TenantId: "tenant-a", Caller: &contractv1.Caller{Kind: contractv1.CallerKind_CALLER_KIND_PLUGIN, Id: "cn.vastplan.platform.configuration.plugin-settings"}}
	result, raw, err := handler(context.Background(), nil, coordinator, payload)
	if err != nil || result.GetStatus() != contractv1.CallResult_STATUS_OK {
		t.Fatalf("有效协调器调用失败: result=%+v err=%v", result, err)
	}
	var observation configurationv1.Observation
	if json.Unmarshal(raw, &observation) != nil || observation.Candidate == nil || observation.Candidate.Status != configurationv1.StatusPrepared {
		t.Fatalf("controller observation 错误: %+v", observation)
	}
}
